certora/specs/Delegation.spec

```
@@ -0,0 +1,58 @@
+import "ERC20.spec";
+
+ghost mathint sumOfVotingPower {
+    init_state axiom sumOfVotingPower == 0;
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sstore (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] uint256 newValue (uint256 oldValue) {
+    sumOfVotingPower = sumOfVotingPower - oldValue + newValue;
+}
+
+// Check that zero address has no voting power assuming that zero address can't make transactions.
+invariant zeroAddressNoVotingPower()
+    delegatee(0x0) == 0x0 && delegatedVotingPower(0x0) == 0
+    { preserved with (env e) { require e.msg.sender != 0; } }
+
+// Check that the voting power plus the virtual voting power of address zero is equal to the total supply of tokens.
+invariant totalSupplyIsSumOfVirtualVotingPower()
+    to_mathint(totalSupply()) == sumOfVotingPower + currentContract._zeroVirtualVotingPower
+    {
+      preserved {
+          // Safe requires because the proxy contract should be initialized right after construction.
+          require totalSupply() == 0;
+          require sumOfVotingPower == 0;
+
+          requireInvariant totalSupplyIsSumOfBalances();
+          requireInvariant zeroAddressNoVotingPower();
+      }
+    }
+
+function isTotalSupplyGTEqSumOfVotingPower() returns bool {
+    requireInvariant totalSupplyIsSumOfVirtualVotingPower();
+    return totalSupply() >= sumOfVotingPower;
+}
+
+// Check that the total supply of tokens is greater than or equal to the sum of voting power.
+rule totalSupplyGTEqSumOfVotingPower {
+    assert isTotalSupplyGTEqSumOfVotingPower();
+}
+
+// Check that users can delegate their voting power.
+rule delegatingUpdatesVotingPower(env e, address newDelegatee) {
+    requireInvariant zeroAddressNoVotingPower();
+    assert isTotalSupplyGTEqSumOfVotingPower();
+
+    address oldDelegatee = delegatee(e.msg.sender);
+
+    mathint delegatedVotingPowerBeforeOfNewDelegatee = delegatedVotingPower(newDelegatee);
```

exhaustivity goes against readability here
```suggestion
    mathint delegatedVotingPowerBefore = delegatedVotingPower(newDelegatee);
```

---

Fixed in https://github.com/morpho-org/morpho-token-upgradeable/pull/92/commits/83b19c47544df212efcc862f76efffcee4c4a140.